File: exercise/java/day23/src/test/java/ControlSystemTest.java
import core.control.ControlSystem;
import core.control.SleighAction;
import core.control.SleighEngineStatus;
import core.control.SleighNotStartedException;
import core.control.ReindeersNeedRestException;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.io.ByteArrayOutputStream;
import java.io.PrintStream;

import static org.junit.jupiter.api.Assertions.*;

class TestControlSystem {
    private final ByteArrayOutputStream outputStreamCaptor = new ByteArrayOutputStream();

    @BeforeEach
    public void setUp() {
        System.setOut(new PrintStream(outputStreamCaptor));
    }

    @Test
    void testStart() {
        // The system has been started
        ControlSystem controlSystem = new ControlSystem();
        controlSystem.action = SleighAction.FLYING;
        controlSystem.status = SleighEngineStatus.OFF;
        controlSystem.startSystem();
        assertTrue(controlSystem.status == SleighEngineStatus.ON);
        assertEquals("Starting the sleigh...\r\n" + "System ready.",
                outputStreamCaptor.toString().trim());
    }

    @Test
    void testAscend() throws ReindeersNeedRestException, SleighNotStartedException {
        ControlSystem controlSystem = new ControlSystem();
        controlSystem.startSystem();
        controlSystem.ascend();
        assertTrue(controlSystem.action == SleighAction.FLYING);
        assertEquals(
                """
                        Starting the sleigh...\r
                        System ready.\r
                        Ascending...""",
                outputStreamCaptor.toString().trim());
    }

    @Test
    void testDescend() throws ReindeersNeedRestException, SleighNotStartedException {
        ControlSystem controlSystem = new ControlSystem();
        controlSystem.startSystem();
        controlSystem.ascend();
        controlSystem.descend();
        assertTrue(controlSystem.action == SleighAction.HOVERING);
        assertEquals(
                """
                        Starting the sleigh...\r
                        System ready.\r
                        Ascending...\r
                        Descending...""",
                outputStreamCaptor.toString().trim());
    }

    @Test
    void testPark() {
        // That would be nice to test it
        assertNotNull(new ControlSystem());
    }
}

File: exercise/kotlin/day23/src/test/kotlin/ControlSystemTest.kt
import core.control.ControlSystem
import core.control.SleighAction
import core.control.SleighEngineStatus
import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import java.io.ByteArrayOutputStream
import java.io.PrintStream

class ControlSystemTest : StringSpec({
    val outputStreamCaptor = ByteArrayOutputStream()

    beforeTest {
        System.setOut(PrintStream(outputStreamCaptor))
    }

    afterTest {
        System.setOut(System.out)
    }

    "testStart" {
        // The system has been started
        val controlSystem = ControlSystem()
        controlSystem.action = SleighAction.FLYING
        controlSystem.status = SleighEngineStatus.OFF
        controlSystem.startSystem()
        controlSystem.status shouldBe SleighEngineStatus.ON
        outputStreamCaptor.toString().trim() shouldBe "Starting the sleigh...\nSystem ready."
    }

    "testAscend" {
        val controlSystem = ControlSystem()
        controlSystem.startSystem()
        controlSystem.ascend()
        controlSystem.action shouldBe SleighAction.FLYING
        outputStreamCaptor.toString().trim() shouldBe "Starting the sleigh...\nSystem ready.\nAscending..."
    }

    "testDescend" {
        val controlSystem = ControlSystem()
        controlSystem.startSystem()
        controlSystem.ascend()
        controlSystem.descend()
        controlSystem.action shouldBe SleighAction.HOVERING
        outputStreamCaptor.toString()
            .trim() shouldBe "Starting the sleigh...\nSystem ready.\nAscending...\nDescending..."
    }

    "testPark" {
        val controlSystem = ControlSystem()
        controlSystem.startSystem()
        controlSystem.park()
        controlSystem.action shouldBe SleighAction.PARKED
    }
})